homeassistant/components/template/__init__.py

```diff
@@ -1,60 +1,109 @@
 """The template component."""
 
-from itertools import chain
+import asyncio
 import logging
 
-from homeassistant.const import MATCH_ALL
+from homeassistant import config as conf_util
+from homeassistant.const import SERVICE_RELOAD
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_per_platform
+from homeassistant.loader import async_get_integration
+
+from .const import (
+    DOMAIN,
+    ENTITIES_STORAGE_KEY,
+    EVENT_TEMPLATE_RELOADED,
+    PLATFORM_STORAGE_KEY,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 
-def initialise_templates(hass, templates, attribute_templates=None):
-    """Initialise templates and attribute templates."""
-    if attribute_templates is None:
-        attribute_templates = {}
-    for template in chain(templates.values(), attribute_templates.values()):
-        if template is None:
-            continue
-        template.hass = hass
+async def _async_setup_reload_service(hass):
+    if hass.services.has_service(DOMAIN, SERVICE_RELOAD):
+        return
 
+    async def _reload_config(call):
+        """Reload the template platform config."""
+        try:
+            unprocessed_conf = await conf_util.async_hass_config_yaml(hass)
+        except HomeAssistantError as err:
+            _LOGGER.error(err)
+            return
 
-def extract_entities(
-    device_name, device_type, manual_entity_ids, templates, attribute_templates=None
-):
-    """Extract entity ids from templates and attribute templates."""
-    if attribute_templates is None:
-        attribute_templates = {}
-    entity_ids = set()
-    if manual_entity_ids is None:
-        invalid_templates = []
-        for template_name, template in chain(
-            templates.items(), attribute_templates.items()
-        ):
-            if template is None:
-                continue
+        remove_tasks = []
+        add_tasks = []
+
+        for platform_setup in hass.data[PLATFORM_STORAGE_KEY].values():
+            platform, _ = platform_setup
 
-            template_entity_ids = template.extract_entities()
-
-            if template_entity_ids != MATCH_ALL:
-                entity_ids |= set(template_entity_ids)
-            else:
-                invalid_templates.append(template_name.replace("_template", ""))
-
-        entity_ids = list(entity_ids)
-
-        if invalid_templates:
-            if not entity_ids:
-                entity_ids = MATCH_ALL
-            _LOGGER.warning(
-                "Template %s '%s' has no entity ids configured to track nor"
-                " were we able to extract the entities to track from the %s "
-                "template(s). This entity will only be able to be updated "
-                "manually",
-                device_type,
-                device_name,
-                ", ".join(invalid_templates),
+            old_entity_ids = [
```

Comment:
It would be much cleaner to make `entity_platform` know that the entity is provided by `template` instead of this and have something like `async_reset("template")`  

Maybe it already does and I missed it?

Reply:
After re-testing with a clean branch `async_reset` works as expected.  I think it might have been caused by some other changes I had in the integration branch.  Sorry for the trip down the rabbit hole.